src/lib.rs

```diff
@@ -155,7 +173,10 @@ const MAX_RECURSION: usize = 64;
 pub struct RegexBuilder(RegexOptions);
 
 /// A compiled regular expression.
-pub struct Regex(RegexImpl);
+pub struct Regex {
+    inner: RegexImpl,
+    group_names: Arc<NamedGroups>,
```

Comment:
Why the `Arc`?

Edit: Ah, I see, the regex crate uses an `Arc` too. But it actually makes use of it, by `clone` ing the arc in the `captures` method: https://docs.rs/regex/1.3.9/src/regex/re_unicode.rs.html#315

Can you:

1. Change `Captures`'s `names` to use an `Arc` too
2. Change your implementation of `captures_from_pos` to use `names: self.group_names.clone(),` instead? That will save a bunch of unnecessary `String` allocations.

Reply:
Done. I also change the field name to `named_groups` everywhere to be consistent with the existing code.